networking_aci/plugins/ml2/drivers/mech_aci/allocations_manager.py

```diff
@@ -226,46 +232,110 @@ def _release_vxlan_segment(self, network, host_config, level, segment):
+                              network_id, segment_physnet, ", ".join(n.id for n in existing_segments),
+                              ", ".join(n.segmentation_id for n in existing_segments))
+                segment = existing_segments[0]
+                if is_access:
+                    # make sure the segment id is from the right pool
+                    if segment.segmentation_id not in access_id_pool:
+                        raise NetworkHasBoundTrunkPorts(network_id=network_id, segment_id=segment.id,
+                                                        segmentation_id=segment.segmentation_id)
+                else:
+                    # make sure the segment id matches, else it's an affinity error
+                    if segment.segmentation_id != segmentation_id:
+                        raise NetworkUsesDifferentTrunkId(network_id=network_id,
+                                                          segmentation_id=segment.segmentation_id)
+                return segment
+
+            # 2. sanity checks
+            if is_access:
+                # 3. for access mode: check that no other network has bound this in host mode
+                host_segments = self.db.get_hosts_on_physnet(ctx, segment_physnet, level=1,
+                                                             with_segment=True, with_segmentation=True)
+                for far_host, far_segment_id, far_segmentation_id in host_segments:
+                    if far_host in hostgroup['hosts'] and far_segmentation_id in access_id_pool:
+                        raise HostAlreadyHasAccessBinding(host=far_host, segmentation_id=far_segmentation_id,
+                                                          segment_id=far_segment_id)
+            else:
+                # for trunk mode: check segmentation_id is not already in use in physnet
+                existing_segments = (session.query(ml2_models.NetworkSegment)
+                                     .filter_by(segmentation_id=segmentation_id, physical_network=segment_physnet,
+                                                segment_index=level, network_type=segment_type)
+                                     .all())
+                if existing_segments:
+                    raise TrunkSegmentIdAlreadyInUse(segmentation_id=segmentation_id,
+                                                     segment_id=existing_segments[0].id)
+
+            # 3. no segment exists, allocate one
+            if is_access:
+                # find a free vlan id from the pool
+                physnet_segments = (session.query(ml2_models.NetworkSegment)
+                                    .filter_by(physical_network=segment_physnet)
+                                    .all())
+                used_ids = set(n.segmentation_id for n in physnet_segments)
+                possible_ids = access_id_pool - used_ids
+                if not possible_ids:
+                    raise NoAllocationFoundInMaximumAllowedAttempts()
```

Comment:
Maybe better exception that the pool is exhausted? Or some details on which pool or something?